src/backend/vulkan/src/lib.rs

```diff
@@ -327,7 +327,21 @@ impl hal::Instance<Backend> for Instance {
             .application_version(version)
             .engine_name(CStr::from_bytes_with_nul(b"gfx-rs\0").unwrap())
             .engine_version(1)
-            .api_version(vk::make_version(1, 0, 0));
+            .api_version(
+                // Pick the latest version of Vulkan available.
+                match entry
+                    .try_enumerate_instance_version()
+                    // Ignore the possible `VK_ERROR_OUT_OF_HOST_MEMORY`.
+                    .unwrap()
+                {
+                    // Vulkan 1.1+
+                    Some(version) => {
+                        vk::make_version(vk::version_major(version), vk::version_minor(version), 0)
```

Comment:
We have code that tracks the maintenance version, and other code that checks for it. Both currently assume 1.0 baseline, and they obviously need to be adjusted if running on 1.1. I think what we'll do is only patch the first half: code that determines what maintenance level to request would associate Vulkan-1.1 with level 3, and Vulkan-1.1 with level 4. The code that checks for it would then be unchanged.

Reply:
Is it an error if I request maintenence1 and version 1.1? If that's the case we should also try and tackle the extension dependency graph issue in this PR as well.

Is it fair to assume that we want to always enable the highest maintenance level available level available to us (for our given api version)?

Reply:
> Is it an error if I request maintenence1 and version 1.1?

It is technically deprecated in Vk1.1, so there is no guarantee it's even going to be exposed in Vk1.1 drivers. I don't think we should try requesting it.

> Are you suggesting using maintenance_level: u8 as a value to represent "api 1.0, maintenance 1, 2, 3, api 1.1, and 1.2, etc"?

For all the code past initialization, `maintenance_level` defines what we can and what we can't ask the Vulkan driver. So this code can probably for the most part not care about the instance version. On Instances 1.0 and 1.1 we could just set the `maintenance_level = 3` (without requesting the actual levels) and the rest of the code will be fine.

Reply:
I'm working on a little bit of code that will resolve the requested extensions list given an api version, the extensions dependencies, and the deprecation state. The goal is to make it straightforward to add more extensions later.

Reply:
@kvark I've pushed what I came up with (here's the specific commit https://github.com/gfx-rs/gfx/pull/3624/commits/c749e50d99f0518b3423ccaa1a3f4fd3e5e9263c). This is starting to smell like "overengineering," so please sanity check me on whether or not it is indeed an improvement!

Reply:
@kvark I took a look at a couple Vulkan engines and it seems most just flatten dependencies and hard code the list.

Some examples:
- The Forge: https://github.com/ConfettiFX/The-Forge/blob/2edbb394883b971ba30498f9175c00bf834db9c6/Common_3/Renderer/Vulkan/Vulkan.cpp#L225
- Diligent: https://github.com/DiligentGraphics/DiligentCore/blob/fa345f59de694ac249c14e987032ef3e341faaa7/Graphics/GraphicsEngineVulkan/src/EngineFactoryVk.cpp#L240

I'm tempted to remove this "extension resolver" code and just hard code the list like before, thoughts?